myst_nbs/case_studies/wrapping_jax_function.myst.md

```diff
@@ -0,0 +1,775 @@
+:::
+
+For a better understanding into {class}`~aesara.graph.basic.Op` JAX conversions, we recommend you check Aesara's [Adding JAX and Numba support for Ops guide](https://aesara.readthedocs.io/en/latest/extending/creating_a_numba_jax_op.html?highlight=JAX)
+
+We can test that our conversion function is working properly by compiling a {func}`aesara.function` with `mode="JAX"`:
+
+```{code-cell} ipython3
+out = hmm_logp_op(
+    emission_observed,
+    emission_signal_true,
+    emission_noise_true,
+    logp_initial_state_true,
+    logp_transition_true,
+)
+jax_fn = aesara.function(inputs=[], outputs=out, mode="JAX")
+jax_fn()
+```
+
+We can also compile a JAX function that computes the log probability of each variable in our PyMC model, similar to {meth}`~pymc.Model.point_logps`
+
+```{code-cell} ipython3
+inputs = model.value_vars
+outputs = model.logpt(sum=False)
+model_logp_jax_fn = aesara.function(inputs, outputs, mode="JAX")
+```
+
+```{code-cell} ipython3
+model_logp_jax_fn(*initial_point.values())
+```
+
+Note that we could have added an equally simple function to convert our `HMMLogpGradOp`, in case we wanted to convert Aesara gradient graphs to JAX. In our case, we don't need to do this because we will rely on JAX `grad` function (or more precisely, NumPyro will rely on it) to obtain these again from our compiled JAX function.
+
+We include a {ref}`aesara_vs_jax <short discussion>` at the end of this document, to help you better understand the trade-offs between writting grahs in Aesara vs JAX, and when you migh want to use one or the other.
+
++++
+
+## Sampling with Numpyro
+
++++
+
+Now that we know our model logp can be entirely compiled to JAX, we can use the handy {mod}`pymc.sampling_jax` module to sample our model using the pure JAX sampler implemented in NumPyro.
+
+```{code-cell} ipython3
+with model:
+    idata_numpyro = pm.sampling_jax.sample_numpyro_nuts(chains=2)
+```
+
+```{code-cell} ipython3
+az.plot_trace(idata_numpyro);
+```
+
+```{code-cell} ipython3
+az.plot_posterior(idata_numpyro, ref_val=true_values);
+```
+
+As expected, sampling results look pretty similar! 
+
+Depending on the model and computer architecture you are using, a pure JAX sampler can provide considerable speedups.
+
++++
+
+## Bonus: Using a single Op to wrap the value and gradients at once
+
++++
+
+We had to create two {class}`~aesara.graph.basic.Op`s, one for the function we cared about and a separate one for its gradients. However, JAX provides a `value_and_grad` utility that can return both the value of a function and its derivatives. This means we can get away with a single {class}`~aesara.graph.basic.Op` if we are clever about it.
+
+It also means we can (potentially) save memory and reuse computation that is shared between the function and its gradients. This may be relevant when working with very large JAX functions.
```

Comment:
```suggestion
Doing this can (potentially) save memory and reuse computation that is shared between the function and its gradients. This may be relevant when working with very large JAX functions.
```